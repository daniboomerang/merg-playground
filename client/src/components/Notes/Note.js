import React, { useRef, useState } from 'react';
import useOnclickOutside from 'react-cool-onclickoutside';
import styled from 'styled-components';
import { CloseCircleTwoTone } from '@ant-design/icons';
import { Affix, Input, message } from 'antd';
import { blue } from '@ant-design/colors';

const MAX_CHARACTERS_PER_LINE = 77;

const NoteContent = styled.div`
  grid-template-columns: 1fr auto;
  justify-content: space-between;
  grid-column-gap: 50px;
  align-items: start;
  display: grid;
`;

const Actions = styled.div.attrs(({ isHovered }) => ({
  visibility: !isHovered ? 'hidden' : 'visible',
}))`
  visibility:  ${(props) => props.visibility};
`;

const NoteTxt = styled.div`
  white-space: pre-line;
`;

const { TextArea } = Input;

const Note = ({
  id, note, onUpdateNote, onDeleteNote, parentScrollableContainer,
}) => {
  const ref = useRef();

  // State hooks
  const [updateOngoing, setUpdateOngoing] = useState(false);
  const [newNote, setNewNote] = useState(note);
  const [isHovered, setIsHovered] = useState(false);

  // Handlers
  const handleUpdateOngoing = () => setUpdateOngoing(true);

  const handleOnChangeNote = (e) => setNewNote(e.target.value);

  const handleUpdateNote = (e) => {
    if ((!e && newNote !== '') || (e && newNote && e.key === 'Enter' && !e.shiftKey)) {
      onUpdateNote({ variables: { _id: id, content: newNote } })
        .catch((updateNoteMutationError) => {
          message.error(updateNoteMutationError);
        }).finally(() => {
          setUpdateOngoing(false);
        });
    }
  };

  const handleDeleteNote = (e) => {
    e.stopPropagation();

    onDeleteNote({ variables: { _id: id }, refetchQueries: ['NotesQuery'] }).then(() => {
      message.success('Note deleted');
    }).catch((deleteNoteMutationError) => {
      message.error(deleteNoteMutationError);
    });
  };

  const handleOnMouseOver = () => setIsHovered(true);
  const handleOnMouseLeave = () => setIsHovered(false);

  useOnclickOutside(ref, () => {
    if (updateOngoing) {
      handleUpdateNote();
    }
  });

  return (
    // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
    <NoteContent
      onClick={handleUpdateOngoing}
      onMouseOver={handleOnMouseOver}
      onMouseLeave={handleOnMouseLeave}
    >
      <div ref={ref}>
        {updateOngoing ? (
          <TextArea
            rows={Math.round(newNote.length / MAX_CHARACTERS_PER_LINE) + 1 || 1}
            value={newNote}
            onChange={handleOnChangeNote}
            onKeyDown={handleUpdateNote}
          />

        ) : (
          <NoteTxt>{ newNote }</NoteTxt>
        )}
      </div>
      <Affix target={() => parentScrollableContainer} offsetTop={25}>
        <Actions isHovered={isHovered}>
          <CloseCircleTwoTone twoToneColor={blue.primary} onClick={handleDeleteNote} style={{ fontSize: '20px' }} />
        </Actions>
      </Affix>
    </NoteContent>
  );
};

export default Note;
