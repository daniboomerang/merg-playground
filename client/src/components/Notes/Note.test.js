import React from 'react';
import { mount } from 'enzyme';
import Note from './Note';

describe('Note', () => {
  const id = 'noteId';
  const note = 'This is a note';
  const updateNoteMutation = jest.fn();
  const deleteNoteMutation = jest.fn();
  const listScrollableContainer = null;
  const wrapper = mount(
    <Note
      id={id}
      note={note}
      onUpdateNote={updateNoteMutation}
      onDeleteNote={deleteNoteMutation}
      parentScrollableContainer={listScrollableContainer}
    />,
  );

  it('should render the Note component correctly', () => {
    expect(wrapper.text()).toBe(note);
    expect(wrapper).toMatchSnapshot();
  });
});
