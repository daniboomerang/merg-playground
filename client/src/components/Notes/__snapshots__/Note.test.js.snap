// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should render the Note component correctly 1`] = `
<Note
  id="noteId"
  note="This is a note"
  onDeleteNote={[MockFunction]}
  onUpdateNote={[MockFunction]}
  parentScrollableContainer={null}
>
  <styled.div
    onClick={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="sc-AxjAm hDmEQb"
      onClick={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <div>
        <styled.div>
          <div
            className="sc-AxiKw cfcBpt"
          >
            This is a note
          </div>
        </styled.div>
      </div>
      <Affix
        offsetTop={25}
        target={[Function]}
      >
        <ResizeObserver
          onResize={[Function]}
        >
          <div>
            <div
              className=""
            >
              <ResizeObserver
                onResize={[Function]}
              >
                <styled.div
                  isHovered={false}
                >
                  <div
                    className="sc-AxirZ ePIktc"
                    visibility="hidden"
                  >
                    <ForwardRef(CloseCircleTwoTone)
                      onClick={[Function]}
                      style={
                        Object {
                          "fontSize": "20px",
                        }
                      }
                      twoToneColor="#1890ff"
                    >
                      <AntdIcon
                        icon={
                          Object {
                            "icon": [Function],
                            "name": "close-circle",
                            "theme": "twotone",
                          }
                        }
                        onClick={[Function]}
                        style={
                          Object {
                            "fontSize": "20px",
                          }
                        }
                        twoToneColor="#1890ff"
                      >
                        <span
                          aria-label="close-circle"
                          className="anticon anticon-close-circle"
                          onClick={[Function]}
                          role="img"
                          style={
                            Object {
                              "fontSize": "20px",
                            }
                          }
                          tabIndex={-1}
                        >
                          <IconReact
                            className=""
                            icon={
                              Object {
                                "icon": [Function],
                                "name": "close-circle",
                                "theme": "twotone",
                              }
                            }
                            primaryColor="#1890ff"
                          >
                            <svg
                              aria-hidden="true"
                              className=""
                              data-icon="close-circle"
                              fill="currentColor"
                              focusable="false"
                              height="1em"
                              key="svg-close-circle"
                              viewBox="64 64 896 896"
                              width="1em"
                            >
                              <path
                                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                                fill="#1890ff"
                                key="svg-close-circle-svg-0"
                              />
                              <path
                                d="M512 140c-205.4 0-372 166.6-372 372s166.6 372 372 372 372-166.6 372-372-166.6-372-372-372zm171.8 527.1c1.2 1.5 1.9 3.3 1.9 5.2 0 4.5-3.6 8-8 8l-66-.3-99.3-118.4-99.3 118.5-66.1.3c-4.4 0-8-3.6-8-8 0-1.9.7-3.7 1.9-5.2L471 512.3l-130.1-155a8.32 8.32 0 01-1.9-5.2c0-4.5 3.6-8 8-8l66.1.3 99.3 118.4 99.4-118.5 66-.3c4.4 0 8 3.6 8 8 0 1.9-.6 3.8-1.8 5.2l-130.1 155 129.9 154.9z"
                                fill="#e6f7ff"
                                key="svg-close-circle-svg-1"
                              />
                              <path
                                d="M685.8 352c0-4.4-3.6-8-8-8l-66 .3-99.4 118.5-99.3-118.4-66.1-.3c-4.4 0-8 3.5-8 8 0 1.9.7 3.7 1.9 5.2l130.1 155-130.1 154.9a8.32 8.32 0 00-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3 99.3-118.5L611.7 680l66 .3c4.4 0 8-3.5 8-8 0-1.9-.7-3.7-1.9-5.2L553.9 512.2l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z"
                                fill="#1890ff"
                                key="svg-close-circle-svg-2"
                              />
                            </svg>
                          </IconReact>
                        </span>
                      </AntdIcon>
                    </ForwardRef(CloseCircleTwoTone)>
                  </div>
                </styled.div>
              </ResizeObserver>
            </div>
          </div>
        </ResizeObserver>
      </Affix>
    </div>
  </styled.div>
</Note>
`;
